Detect lockdown

fbs opened the pull request:
The first 4 commits of this PR add map and probe type detection to `BPFfeature` and remove some of the boiler plate. To detect a helper adding ` EMIT_HELPER_TEST(override_return, BPF_PROG_TYPE_KPROBE);` should now be enough (after update bpffeature.h).

The last commit adds lock down detection. The detection is based on these two commits:

- https://github.com/torvalds/linux/commit/a94549dd87f5ea4ca50fee493df08a2dc6256b53
- https://github.com/torvalds/linux/commit/9d1f8be5cf42b497a3bddf1d523f2bb142e9318c

If both kprobes and probe_read support is it is likely that lockdown is in effect. It could still be a false positive but in any case the system is not suitable to run bpftrace.

My only test system is an ubuntu 19.10 workstation. For some reason bpf appears to be fully blocked on it, even map creation fails. This doesn't appear to be part of the original patches, but it only happens when secure boot is enabled. It would be good to have more people test this.

```
sudo strace -e bpf -c  bpftrace  --info
The system appears to be in lockdown mode (kprobes and probe_reads are blocked).
Lockdown mode blocks parts of BPF which makes it impossible for bpftrace to function.
Please see https://github.com/iovisor/bpftrace/blob/master/INSTALL.md#disable-lockdown
for more details on lockdown and how to disable it.
Keep in mind that detection is not entirely accurate, there are other
security mechanisms that can cause this too

Kernel helpers
  get_current_cgroup_id: no
  send_signal: no
  override_return: no

Kernel features
  Instruction limit: -1
  Loop support: no

Map types
  hash: no
  percpu hash: no
  array: no
  percpu array: no
  stack_trace: no
  perf_event_array: no

Probe types
  kprobe: no
  tracepoint: no
  perf_event: no

% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
100,00    0,000135           3        40        40 bpf
------ ----------- ----------- --------- --------- ----------------
100.00    0,000135                    40        40 total
```

Relates to: #925

fbs commented on INSTALL.md:
````diff
@@ -250,3 +251,14 @@ To test that the build works, you can try running the test suite, and a one-line
 ```
+
+# Disable Lockdown
+
````
not sure what to put here, other than disabling secure boot in the bios

tyhicks replied:
> not sure what to put here, other than disabling secure boot in the bios

Disabling secure boot in the bios is one option. Another option is `sudo mokutil --disable-validation`, rebooting, and then following the prompts to disable secure boot verification in `shim`.

fbs commented:
ah found the issue with ubuntu, they ship this (bug):
```
commit 2a68c65abae66d28e2acb3245cb156ae2ea6eb1d
Author: David Howells <dhowells@redhat.com>
Date:   Mon Mar 25 15:09:50 2019 -0700

    UBUNTU: SAUCE: (efi-lockdown) bpf: Restrict kernel image access functions when the kernel is locked down

    There are some bpf functions can be used to read kernel memory:
    bpf_probe_read, bpf_probe_write_user and bpf_trace_printk.  These allow
    private keys in kernel memory (e.g. the hibernation image signing key) to
    be read by an eBPF program and kernel memory to be altered without
    restriction.

    Completely prohibit the use of BPF when the kernel is locked down.

diff --git a/kernel/bpf/syscall.c b/kernel/bpf/syscall.c
index 84470d1480aa..d6f3b787d90d 100644
--- a/kernel/bpf/syscall.c
+++ b/kernel/bpf/syscall.c
@@ -2593,6 +2593,9 @@ SYSCALL_DEFINE3(bpf, int, cmd, union bpf_attr __user *, uattr, unsigned int, siz
        if (sysctl_unprivileged_bpf_disabled && !capable(CAP_SYS_ADMIN))
                return -EPERM;

+       if (kernel_is_locked_down("BPF"))
+               return -EPERM;
+
```

git://git.launchpad.net/~ubuntu-kernel/ubuntu/+source/linux/+git/disco

brendangregg commented:
For detection there's also

```
# mokutil --sb-state
SecureBoot disabled
```

If that says enabled on Ubuntu, then lockdown is also on.

tyhicks commented:
> ah found the issue with ubuntu, they ship this (bug):

This is currently being relaxed in Ubuntu to allow `bpf(2)` but `bpf_probe_read` and `bpf_probe_read_str` will still be restricted in lockdown mode. See the following bug for details:

 https://bugs.launchpad.net/bugs/1863234

I know that this still isn't the perfect solution for users of bpf but I wanted to provide the updated status here. Hope it helps some.

jeremycline commented on src/main.cpp:
```diff
@@ -78,6 +78,33 @@ void usage()
+/*
+  Return true and print a warning message if lockdown is enabled.
+*/
+static bool warn_on_lockdown(BPFfeature* feature)
```
For what it's worth I sent https://lore.kernel.org/lkml/20200220151738.1492852-1-jcline@redhat.com/ so once it's merged you should be able to read /sys/kernel/security/lockdown for the current status as an unprivileged user.

fbs replied:
Ah nice. Bpftrace only supports running as root at the moment so we don't  really need it yet but good to have :)

fbs commented:
I've pushed a new version which makes use the `/sys/kernel/security/lockdown` file. 

Ubuntu is still a special case for now. We can probably remove the exception around the time 20.04 comes out as most users will probably have updated (19.10 latest or 20.04).

With this new approach the changes to bpffeature aren't required anymore, but I've kept them in as I think that they're good to have anyway.

I still need to update the INSTALL notes though

danobi commented on src/lockdown.cpp:
```diff
@@ -0,0 +1,87 @@
+  // Format: none [integrity] confidentiality
+  while (!file.fail())
+  {
+    std::string line;
```
This loop was a little confusing since the temp var was named `line` (also the entire `std::*stream` API is confusing). Maybe use `split_string()` or name this `entry`.

fbs replied:
ill rename it to field an add an extra comment.

Not a fan of `*stream` either but I think this works ok in this case.

fbs commented on src/bpffeature.h:
```diff
@@ -1,43 +1,32 @@
-  bool has_override_return_;
+  std::unique_ptr<bool> has_send_signal_;
+  std::unique_ptr<bool> has_get_current_cgroup_id_;
+  std::unique_ptr<bool> has_override_return_;
```
> Seems like an inappropriate use of unique_ptr. I know we have our build set to c++14, but how about bumping to c++17 and using `std::optional`?

Agreed it is a bit hacky but should work ok for this purpose. Switching to c++17 means we have to have gcc >= 7, not sure how many issues that it going to cause (e.g. ubuntu xenial would be an issue)

danobi replied:
Agreed, won't block PR on this.

As a side note, I think you can install newer GCCs from this PPA: https://launchpad.net/~ubuntu-toolchain-r/+archive/ubuntu/test . GCC 7 has been out for almost 3 years now, I think it would be reasonable to depend on it be available.

fbs replied:
created an issue to look at c++17 (#1170 )

fbs commented:
@danobi  I've updated it  to:

- delete the copy/move constructor/assignment as I don't think they make sense with this unique_ptr hack
- updated the macro to define both the variable and the helper function

fbs commented:
I've updated the wording too, @brendangregg  can you take a look at it?

fbs commented:
c++17 support has been enabled but seems to cause some issues for people building on ubunt 16.04. If we decide to keep 17 ill create a follow up to convert this from unique_ptr to std::optional
